src/network/message.rs

```diff
@@ -37,21 +37,64 @@ use consensus::encode::MAX_VEC_SIZE;
 #[derive(PartialEq, Eq, Clone, Debug)]
 pub struct CommandString(Cow<'static, str>);
 
+impl CommandString {
+    /// Convert from various string types into a [CommandString].
+    ///
+    /// Supported types are:
+    /// - `&'static str`
+    /// - `String`
+    /// - `&str`
+    ///
+    /// Returns an empty error if and only if the string is
+    /// larger than 12 characters in length.
+    pub fn from<S: Into<Cow<'static, str>>>(s: S) -> Result<CommandString, ()> {
+        let cow = s.into();
+        if cow.as_ref().len() > 12 {
+            Err(())
+        } else {
+            Ok(CommandString(cow))
+        }
+    }
+}
+
 impl fmt::Display for CommandString {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.write_str(self.0.as_ref())
     }
 }
 
+impl str::FromStr for CommandString {
+    type Err = ();
+    fn from_str(s: &str) -> Result<CommandString, ()> {
+        // Don't use CommandString::from to avoid allocating when
+        // it's not needed.
+        if s.len() > 12 {
+            Err(())
+        } else {
+            Ok(CommandString(s.to_owned().into()))
+        }
+    }
+}
+
+#[deprecated(since = "0.26.0", note = "use CommandString::from")]
 impl From<&'static str> for CommandString {
     fn from(f: &'static str) -> Self {
-        CommandString(f.into())
+        if f.len() > 12 {
+            CommandString(f[0..12].into())
+        } else {
+            CommandString(f.into())
+        }
     }
 }
 
+#[deprecated(since = "0.26.0", note = "use CommandString::from")]
```

Comment:
Trying to understand why this and above are depricated. The message will be really surprising in practice: instead of `something.from()` users are asked to use `something::<CommandString>.from()`, but even that is not clear from the message... Because in most cases method is called on the object, without additional qualifiers, and it will be up to the compiler to decide wich implementation to call. 

It seams reasonable just to remove this methods, and nothing bad will happen.

Reply:
Yeah I removed it. I pushed a new commit cleaning up the CommandString API.